backend/src/spotify.py

```diff
@@ -21,119 +26,243 @@
 ]
 
 
+class BearerAuth(requests.auth.AuthBase):
+    def __init__(self, token):
+        self.token = token
+
+    def __call__(self, r):
+        r.headers["authorization"] = "Bearer " + self.token
+        return r
+
+
 class SpotifyClient:
     def __init__(self):
-        client_id = os.getenv("SPOTIFY_CLIENT_ID")
-        client_secret = os.getenv("SPOTIFY_SECRET")
-        redirect_uri = os.getenv("SPOTIPY_REDIRECT_URI")
-
-        self.client = spotipy.Spotify(
-            auth_manager=SpotifyOAuth(
-                scope=scope,
-                client_id=client_id,
-                client_secret=client_secret,
-                redirect_uri=redirect_uri,
-            )
+        self.client_id = os.getenv("SPOTIFY_CLIENT_ID")
+        self.client_secret = os.getenv("SPOTIFY_SECRET")
+        self.redirect_uri = os.getenv("SPOTIFY_REDIRECT_URI")
+        self.scope = scope
+
+    def get_login_query_string(self, state):
+        return urlencode(
+            {
+                "response_type": "code",
+                "client_id": self.client_id,
+                "scope": " ".join(self.scope),
+                "redirect_uri": self.redirect_uri,
+                "state": state,
+            }
         )
 
-    def get_playlists(self, limit=10, offset=0):
-        api_playlists = self.client.current_user_playlists(limit=limit, offset=offset)
-        return api_playlists["items"]
+    def request_access_token(self, code):
+        response = requests.post(
+            url="https://accounts.spotify.com/api/token",
+            data={
+                "code": code,
+                "redirect_uri": self.redirect_uri,
+                "grant_type": "authorization_code",
+            },
+            headers={
+                "content-type": "application/x-www-form-urlencoded",
+            },
+            auth=(self.client_id, self.client_secret),
+        )
+        access_token = response.json()["access_token"]
+        user_info = self.get_current_user(access_token)
 
-    def get_playlist(self, id: str, fields=None):
-        api_playlist = self.client.playlist(playlist_id=id, fields=fields)
-        return api_playlist
+        resp = make_response(redirect("http://localhost:1234/"))
```

Comment:
Watch out for the hardcoded URLs like this here - probably worth extracting to an env variable sooner rather than later

Reply:
Think I do a lot of this in the Ansible ticket. If it's alright I'll leave it there to avoid merge conflicts